File: web/src/dashboards.ts
/**
 * Dashboards: a name, and an ordered list of panels.
 *
 * A panel is a `Query` AST and a picture to draw it as — the third thing in this app built
 * on that one type, after the Explorer and alert rules.
 *
 * The two functions with tests are the two that are quietly wrong when they are wrong:
 * `forWindow`, which decides what a panel actually asks for when the header's range
 * changes, and `toSeries`, which decides which column of a result set is a line.
 */

import { request } from "./api";
import { bucketSeconds, type Query, type ResultSet } from "./query";

export type Viz =
  | { kind: "time_series"; unit?: string }
  | { kind: "stat"; unit?: string; decimals?: number }
  | { kind: "table" }
  | { kind: "gauge"; min: number; max: number; unit?: string }
  | { kind: "alerts" };

export interface Panel {
  id: string;
  title: string;
  /**
   * Absent only for the alerts panel, which reads the control plane.
   *
   * `| undefined` spelled out because this project has `exactOptionalPropertyTypes` on,
   * and a panel built from a kind that takes no query genuinely carries the key with
   * nothing in it.
   */
  query?: Query | undefined;
  viz: Viz;
  /** Columns of twelve. */
  width: number;
  /** Rows of `ROW_HEIGHT`. */
  height: number;
}

export interface Dashboard {
  id: string;
  name: string;
  description: string;
  panels: Panel[];
  created_at: string;
  updated_at: string;
}

export interface DashboardRequest {
  name: string;
  description?: string;
  panels: Panel[];
}

/** How tall one grid row is, in pixels. */
export const ROW_HEIGHT = 110;

export function listDashboards(tenant: string) {
  return request<Dashboard[]>("/api/v1/dashboards", { tenant });
}

export function getDashboard(tenant: string, id: string) {
  return request<Dashboard>(`/api/v1/dashboards/${id}`, { tenant });
}

export function createDashboard(tenant: string, body: DashboardRequest) {
  return request<Dashboard>("/api/v1/dashboards", { method: "POST", body, tenant });
}

export function replaceDashboard(tenant: string, id: string, body: DashboardRequest) {
  return request<Dashboard>(`/api/v1/dashboards/${id}`, {
    method: "PUT",
    body,
    tenant,
  });
}

export function deleteDashboard(tenant: string, id: string) {
  return request<void>(`/api/v1/dashboards/${id}`, { method: "DELETE", tenant });
}

/**
 * What a panel actually asks for, over the range in the header.
 *
 * Two substitutions, and the second is the one that matters:
 *
 * 1. **The window.** Stored absolute, like every `Query` in this product; replaced on
 *    read, like a saved search's. A panel that kept asking about the afternoon it was
 *    created on would be a screenshot.
 *
 * 2. **The bucket.** A panel saved over an hour has five-minute buckets in its
 *    `group_by`. Viewed over thirty days that is 8 640 of them — more than the panel's
 *    own limit, so the chart would silently show the first few hundred and look like a
 *    complete picture of a shorter period. Rewriting the bucket to suit the window is
 *    what makes a dashboard's time picker mean anything, and it is why the thirty-day
 *    acceptance criterion exercises the rollups rather than the raw table: a wide window
 *    asks for wide buckets, and the compiler answers those from the pre-aggregate.
 */
export function forWindow(query: Query, from: Date, to: Date): Query {
  const seconds = bucketSeconds(from.getTime(), to.getTime());

  return {
    ...query,
    time: { start: from.toISOString(), end: to.toISOString() },
    ...(query.group_by
      ? {
          group_by: query.group_by.map((field) =>
            field.field === "time_bucket" ? { field: "time_bucket" as const, seconds } : field,
          ),
        }
      : {}),
  };
}

/** One line on a chart. */
export interface Series {
  /** The labels the query grouped by, joined — or the aggregate's name when there are none. */
  name: string;
  points: { at: number; value: number }[];
}

/**
 * Read a grouped result set as lines.
 *
 * The column layout is the compiler's: `group_by` columns in order, then the aggregations.
 * A time-series panel groups by `time_bucket` and possibly by other things, so column 0
 * is the instant, the last column is the value, and anything between names the line.
 *
 * Rows whose value is null are dropped rather than drawn as zero — `avg()` over a bucket
 * with no samples is null, and a chart that drew it as zero would show an outage as a
 * collapse to the floor, which is a different and much more alarming picture.
 */
export function toSeries(result: ResultSet): Series[] {
  const width = result.columns.length;
  if (width < 2) return [];

  const lines = new Map<string, Series>();

  for (const row of result.rows) {
    const at = Date.parse(String(row[0]).replace(" ", "T") + "Z");
    const raw = row[width - 1];
    const value = typeof raw === "number" ? raw : Number(raw);
    if (Number.isNaN(at) || raw === null || raw === undefined || Number.isNaN(value)) {
      continue;
    }

    const name =
      width === 2
        ? (result.columns[width - 1]?.name ?? "value")
        : row
            .slice(1, width - 1)
            .map((cell) => String(cell ?? ""))
            .join(" · ");

    const line = lines.get(name) ?? { name, points: [] };
    line.points.push({ at, value });
    lines.set(name, line);
  }

  // Sorted by time within each line: the compiler's ORDER BY is the panel's, and a panel
  // that did not ask for one would otherwise draw a polyline that doubles back on itself.
  for (const line of lines.values()) {
    line.points.sort((a, b) => a.at - b.at);
  }

  return [...lines.values()];
}

/**
 * The one number a stat or a gauge shows.
 *
 * The **last** row's aggregate, which for a query grouped by time is the most recent
 * bucket — "what is it now", which is what somebody reads a stat panel for. A mean across
 * the window would be a different question, and one the panel's title would not match.
 *
 * `null` when there is nothing to show, which the panel renders as "no data" rather than
 * as zero: those are different facts and only one of them is reassuring.
 */
export function latest(result: ResultSet): number | null {
  const width = result.columns.length;
  if (width === 0) return null;

  for (let i = result.rows.length - 1; i >= 0; i -= 1) {
    const raw = result.rows[i]?.[width - 1];
    if (raw === null || raw === undefined) continue;
    const value = typeof raw === "number" ? raw : Number(raw);
    if (!Number.isNaN(value)) return value;
  }
  return null;
}

/** A blank panel of a kind, for the editor. */
export function blank(kind: Viz["kind"], id: string, query?: Query): Panel {
  const viz: Viz =
    kind === "gauge"
      ? { kind, min: 0, max: 100 }
      : kind === "stat"
        ? { kind, decimals: 2 }
        : kind === "time_series"
          ? { kind }
          : kind === "table"
            ? { kind }
            : { kind: "alerts" };

  return {
    id,
    title: "",
    ...(kind === "alerts" ? {} : { query }),
    viz,
    width: kind === "alerts" ? 12 : 6,
    height: 2,
  };
}

File: crates/uops-store-pg/src/dashboards.rs
//! Dashboards — SPEC §M4.
//!
//! A panel is `{ query, viz }`: the same `Query` AST the Explorer posts, a saved search
//! stores and an alert rule holds, plus how to draw it. There is no fourth representation
//! of a question anywhere in this product, and a panel is the third thing that proves it —
//! a saved search with a picture on it.
//!
//! # Every panel's query is compiled before the dashboard is stored
//!
//! The same check saving a search and writing a rule make, for a third reason: a panel
//! that cannot be answered renders as an error box on a wall display that nobody is
//! looking at closely, next to nineteen panels that work. Refusing it at save time puts
//! the sentence in front of the person who can fix it.
//!
//! # The window in a panel's query is provenance
//!
//! Same as a saved search, same as a rule. A dashboard is read over the time range in its
//! header, and every panel's window is replaced when it is read — the *span* is not even
//! kept here, because unlike an alert rule ("avg over five minutes") a panel is a picture
//! of whatever range the person is looking at.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uops_core::{ActorId, Error as CoreError, Result, TenantScope};
use uops_query::{Query, ResolvedResources, compile};

use crate::error::map;
use crate::store::PgStore;

/// How a panel draws what its query returned.
///
/// The five SPEC names for v0.1, and deliberately not heatmap, geo map, topology or pie:
/// *"five types built well beats twelve built badly"*, and two of those four depend on M6
/// and M7 anyway.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case", tag = "kind")]
pub enum Viz {
    /// A line per series over the panel's window.
    TimeSeries {
        #[serde(default)]
        unit: String,
    },
    /// One number: the most recent value the query returned.
    Stat {
        #[serde(default)]
        unit: String,
        #[serde(default = "two")]
        decimals: u8,
    },
    /// The rows as they came back. The escape hatch for a question no picture answers.
    Table,
    /// One number against a range somebody chose.
    ///
    /// `min` and `max` are required rather than derived from the data: a gauge whose
    /// bounds move with what it is showing is a gauge that always reads half full.
    Gauge {
        min: f64,
        max: f64,
        #[serde(default)]
        unit: String,
    },
    /// What is firing right now. The one panel with no query of its own — it reads the
    /// alert list, which is the control plane rather than telemetry.
    Alerts,
}

const fn two() -> u8 {
    2
}

impl Viz {
    /// Whether this kind of panel reads telemetry.
    #[must_use]
    pub const fn needs_query(&self) -> bool {
        !matches!(self, Self::Alerts)
    }

    #[must_use]
    pub const fn as_str(&self) -> &'static str {
        match self {
            Self::TimeSeries { .. } => "time_series",
            Self::Stat { .. } => "stat",
            Self::Table => "table",
            Self::Gauge { .. } => "gauge",
            Self::Alerts => "alerts",
        }
    }
}

/// One panel.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Panel {
    /// Stable for the life of the panel, so a client can key a list on it and a layout
    /// change does not remount every chart on the page.
    pub id: String,
    pub title: String,
    /// `None` only for [`Viz::Alerts`] — enforced by [`validate`].
    #[serde(default)]
    pub query: Option<Query>,
    pub viz: Viz,
    /// Columns of twelve. Twelve because it divides by two, three, four and six, which is
    /// every layout somebody actually asks for.
    #[serde(default = "six")]
    pub width: u8,
    /// Rows of a fixed height, decided by the client. Stored because the layout is part of
    /// the dashboard rather than of the browser that drew it last.
    #[serde(default = "two_rows")]
    pub height: u8,
}

const fn six() -> u8 {
    6
}

const fn two_rows() -> u8 {
    2
}

/// A dashboard as stored.
#[derive(Clone, Debug)]
pub struct Dashboard {
    pub id: uuid::Uuid,
    pub tenant_id: uops_core::TenantId,
    pub name: String,
    pub description: String,
    pub panels: Vec<Panel>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// What a caller supplies to create or replace one.
#[derive(Clone, Debug)]
pub struct NewDashboard {
    pub name: String,
    pub description: String,
    pub panels: Vec<Panel>,
}

/// Reject a dashboard that could not be drawn.
///
/// Three things, each of which otherwise becomes an error box on a wall display:
/// a panel whose query the compiler refuses, a telemetry panel with no query at all, and
/// an alert panel carrying one it would ignore.
fn validate(panels: &[Panel], scope: &TenantScope) -> Result<()> {
    for panel in panels {
        match (&panel.query, panel.viz.needs_query()) {
            (Some(query), true) => {
                compile(query, scope, &ResolvedResources::whole_tenant(scope))
                    .map(|_| ())
                    .map_err(|e| {
                        CoreError::Invalid(format!("panel {}: {e}", panel.title.trim()))
                    })?;
            }
            (None, true) => {
                return Err(CoreError::Invalid(format!(
                    "panel {} is a {} and has no query — it would draw nothing",
                    panel.title.trim(),
                    panel.viz.as_str()
                )));
            }
            // An alerts panel with a query attached is a misunderstanding worth
            // correcting rather than silently ignoring: somebody built a query expecting
            // it to filter the alert list, and it would not.
            (Some(_), false) => {
                return Err(CoreError::Invalid(format!(
                    "panel {} shows the alert list and cannot take a query",
                    panel.title.trim()
                )));
            }
            (None, false) => {}
        }
    }

    Ok(())
}

impl PgStore {
    /// Create a dashboard.
    pub async fn create_dashboard(
        &self,
        scope: &TenantScope,
        by: Option<ActorId>,
        new: &NewDashboard,
    ) -> Result<Dashboard> {
        validate(&new.panels, scope)?;
        let panels = serde_json::to_value(&new.panels)?;

        // tenant-exempt: the tenant is the first bound parameter, from the scope.
        let row = sqlx::query!(
            r#"
            INSERT INTO dashboard (tenant_id, name, description, panels, created_by)
            VALUES ($1, $2, $3, $4, $5)
            RETURNING id, tenant_id AS "tenant_id: uops_core::TenantId",
                      name, description, panels, created_at, updated_at
            "#,
            scope.tenant_id() as uops_core::TenantId,
            new.name.trim(),
            new.description,
            panels,
            by as Option<ActorId>,
        )
        .fetch_one(self.pool())
        .await
        .map_err(|e| map("dashboard", new.name.clone(), e))?;

        Ok(Dashboard {
            id: row.id,
            tenant_id: row.tenant_id,
            name: row.name,
            description: row.description,
            panels: serde_json::from_value(row.panels)?,
            created_at: row.created_at,
            updated_at: row.updated_at,
        })
    }

    /// A tenant's dashboards, by name.
    ///
    /// Panels included: a list of dashboards is a handful of rows, and the alternative is
    /// a second request per dashboard to find out how many panels each has.
    pub async fn dashboards(&self, scope: &TenantScope) -> Result<Vec<Dashboard>> {
        // tenant-exempt: the tenant is the only bound parameter, from the scope.
        let rows = sqlx::query!(
            r#"
            SELECT id, tenant_id AS "tenant_id: uops_core::TenantId",
                   name, description, panels, created_at, updated_at
              FROM dashboard
             WHERE tenant_id = $1
             ORDER BY name
            "#,
            scope.tenant_id() as uops_core::TenantId,
        )
        .fetch_all(self.pool())
        .await
        .map_err(|e| map("dashboard", "list".to_owned(), e))?;

        rows.into_iter()
            .map(|row| {
                Ok(Dashboard {
                    id: row.id,
                    tenant_id: row.tenant_id,
                    name: row.name,
                    description: row.description,
                    panels: serde_json::from_value(row.panels)?,
                    created_at: row.created_at,
                    updated_at: row.updated_at,
                })
            })
            .collect()
    }

    /// One dashboard.
    ///
    /// # Errors
    ///
    /// `NotFound` for another tenant's, which is the same answer as for one that does not
    /// exist.
    pub async fn dashboard(&self, scope: &TenantScope, id: uuid::Uuid) -> Result<Dashboard> {
        // tenant-exempt: the tenant is a bound parameter, from the scope.
        let row = sqlx::query!(
            r#"
            SELECT id, tenant_id AS "tenant_id: uops_core::TenantId",
                   name, description, panels, created_at, updated_at
              FROM dashboard
             WHERE tenant_id = $1 AND id = $2
            "#,
            scope.tenant_id() as uops_core::TenantId,
            id,
        )
        .fetch_optional(self.pool())
        .await
        .map_err(|e| map("dashboard", id.to_string(), e))?
        .ok_or(CoreError::NotFound {
            kind: "dashboard",
            id: id.to_string(),
        })?;

        Ok(Dashboard {
            id: row.id,
            tenant_id: row.tenant_id,
            name: row.name,
            description: row.description,
            panels: serde_json::from_value(row.panels)?,
            created_at: row.created_at,
            updated_at: row.updated_at,
        })
    }

    /// Replace a dashboard.
    ///
    /// Whole, like a saved search and an alert rule: a dashboard is a document, and a
    /// partial update of a panel list is a merge nobody can review.
    pub async fn update_dashboard(
        &self,
        scope: &TenantScope,
        id: uuid::Uuid,
        new: &NewDashboard,
    ) -> Result<Dashboard> {
        validate(&new.panels, scope)?;
        let panels = serde_json::to_value(&new.panels)?;

        // tenant-exempt: the tenant is a bound parameter, from the scope.
        let row = sqlx::query!(
            r#"
            UPDATE dashboard
               SET name = $3, description = $4, panels = $5
             WHERE tenant_id = $1 AND id = $2
            RETURNING id, tenant_id AS "tenant_id: uops_core::TenantId",
                      name, description, panels, created_at, updated_at
            "#,
            scope.tenant_id() as uops_core::TenantId,
            id,
            new.name.trim(),
            new.description,
            panels,
        )
        .fetch_optional(self.pool())
        .await
        .map_err(|e| map("dashboard", new.name.clone(), e))?
        .ok_or(CoreError::NotFound {
            kind: "dashboard",
            id: id.to_string(),
        })?;

        Ok(Dashboard {
            id: row.id,
            tenant_id: row.tenant_id,
            name: row.name,
            description: row.description,
            panels: serde_json::from_value(row.panels)?,
            created_at: row.created_at,
            updated_at: row.updated_at,
        })
    }

    /// Delete a dashboard.
    pub async fn delete_dashboard(&self, scope: &TenantScope, id: uuid::Uuid) -> Result<()> {
        // tenant-exempt: the tenant is a bound parameter, from the scope.
        let done = sqlx::query!(
            "DELETE FROM dashboard WHERE tenant_id = $1 AND id = $2",
            scope.tenant_id() as uops_core::TenantId,
            id,
        )
        .execute(self.pool())
        .await
        .map_err(|e| map("dashboard", id.to_string(), e))?;

        if done.rows_affected() == 0 {
            return Err(CoreError::NotFound {
                kind: "dashboard",
                id: id.to_string(),
            });
        }
        Ok(())
    }
}
